_posts/2024-05-27-A-deep-dive-into-faster-semantic-sparse-retrieval-in-OS-2.12.md

```diff
@@ -0,0 +1,144 @@
+1. **For 8.8M documents, neural sparse search benefits from Lucene upgrades more than BM25.** This is because pruning algorithms like WAND and MAXScore in Lucene struggle to efficiently handle the learned sparse weights used by neural search, compared to the term frequencies that BM25 relies on. Additionally, neural sparse search needs to sort query clauses in order to calculate the minimum competitive score. For older Lucene versions, this step is associated with a greater overhead, hindering performance. Thankfully, [this issue has been addressed](https://github.com/apache/lucene/pull/12490) in newer Lucene versions.
+1. **For 1M documents, we observed a noticeable performance improvement for both BM25 and the doc-only mode of neural sparse search.** However, the gains are less pronounced compared to the 8.8M document set. This is because the overall latency for 1M documents is already relatively low, and some fixed costs, like network I/O and tokenizer inference, take a larger proportion of the total processing time. **For the bi-encoder mode, the overall performance improvement is more limited. In this case, the model inference becomes the bottleneck for query throughput.** Because the model is at the limit of processing requests, some search requests are queued for inference.
+
+## Accelerating model inference using GPU
+
+In the previous section, we saw that model inference is the bottleneck for query throughput. Even with the inverted index performance improvements, we won't be able to handle more queries per second without addressing model inference limitations.
+
+The model inference process involves numerous matrix operations, which rely heavily on CPU resources within the cluster nodes. Fortunately, GPUs are specifically designed for this type of workload. Their SIMD (Single Instruction, Multiple Data) architecture and optimizations for matrix operations make them much more efficient for handling pure model inference tasks. This translates to better performance at a lower cost, compared to CPUs. Additionally, techniques like mixed-precision computation and batch inference can further accelerate GPU-based model inference throughput.
+
+We used a SageMaker GPU endpoint to host the neural sparse model. We connected to the model using a [SageMaker connector](https://github.com/zhichao-aws/neural-search/tree/neural_sparse_sagemaker/neural_sparse_sagemaker_example). We chose the **ml.g4dn.xlarge** instance because it is **much cheaper** than the ML node instance (r5.12xlarge) used in the last experiment. After evaluating several popular instances on Amazon SageMaker, we identified the ml.g4dn.xlarge instance as the most cost-effective option for our experiments based on its cost/performance ratio. Detailed results for processing 1 million documents are presented in the following sections.
```

Comment:
```suggestion
We used a SageMaker GPU endpoint to host the neural sparse model. We connected to the model using a [SageMaker connector](https://github.com/zhichao-aws/neural-search/tree/neural_sparse_sagemaker/neural_sparse_sagemaker_example). We chose the **ml.g4dn.xlarge** instance because it is **much cheaper** than the ML node instance (r5.12xlarge) used in the last experiment. After evaluating several popular instances on Amazon SageMaker, we identified the ml.g4dn.xlarge instance as the most cost-effective option for our experiments based on its cost/performance ratio. Detailed results for processing 1M documents are presented in the following sections.
```